feat: base-ragsearch-plugin

alextitonis opened the pull request:
<!-- Use this template by filling in information and copy and pasting relevant items out of the html comments. -->

# Relates to:

# Risks
Low, it's an external plugin

# Background

## What does this PR do?
Adds support for neo4j, users can add documents etc and do rag graph using the provider

## What kind of change is this?

<!--
Bug fixes (non-breaking change which fixes an issue)
Improvements (misc. changes to existing features)
Features (non-breaking change which adds functionality)
Updates (new versions of included code)
-->

<!-- This "Why" section is most relevant if there is no linked issue explaining why. If there is a related issue it might make sense to skip this why section. -->
<!--
## Why are we doing this? Any context or related work?
-->

# Documentation changes needed?

<!--
My changes do not require a change to the project documentation.
My changes require a change to the project documentation.
If a docs change is needed: I have updated the documentation accordingly.
-->

<!-- Please show how you tested the PR. This will really help if the PR needs to be retested, and probably help the PR get merged quicker. -->

# Testing
You can add the provider and try adding some docs and then use the query

## Where should a reviewer start?

## Detailed testing steps

<!--
None, automated tests are fine.
-->

<!--
- As [anon/admin], go to [link]
  - [do action]
  - verify [result]
-->

<!-- If there is a UI change, please include before and after screenshots or videos. This will speed up PRs being merged. It is extra nice to annotate screenshots with arrows or boxes pointing out the differences. -->
<!--
## Screenshots
### Before
### After
-->

<!-- If there is anything about the deploy, please make a note. -->
<!--
# Deploy Notes
-->

<!--  Copy and paste commandline output. -->
<!--
## Database changes
-->

<!--  If there is something more than the automated steps, please specifiy deploy instructions. -->
<!--
## Deployment instructions
-->

<!-- If you are on Discord, please join https://discord.gg/ai16z and state your Discord username here for contribute role and join us in #development-feed -->
<!--
## Discord username

-->

samarth30 commented:
@alextitonis I just reviewed this PR, and I have a suggestion:

This implementation seems more suited to being an Adapter, similar to existing ones like PostgreSQL, Supabase, or SQL Adapters. Could you refactor it into an Adapter with all the relevant functions required for seamless integration?

Additionally, to ensure usability and proper testing, could you include an example demonstrating how this Adapter can be used with an Agent Client for testing purposes? This will help verify the implementation and provide a practical use case.

alextitonis commented:
> @alextitonis I just reviewed this PR, and I have a suggestion:
> 
> This implementation seems more suited to being an Adapter, similar to existing ones like PostgreSQL, Supabase, or SQL Adapters. Could you refactor it into an Adapter with all the relevant functions required for seamless integration?
> 
> Additionally, to ensure usability and proper testing, could you include an example demonstrating how this Adapter can be used with an Agent Client for testing purposes? This will help verify the implementation and provide a practical use case.

It's not meant to be used as a normal database as it's using a graph for storing data, but as an extra layer, that's why i made it like that.
For use case, you can store some memories and draw connections and then use the query to get docs.
For example user X is base.
Chat with X on twitter is on graph, connected to base X
Chat with X on discord is another graph, connected to base
You can query both graphs.
Also will search node by node as well to using the embedding. (not sure if i explained correctly, if not let me know)

samarth30 commented:
Looks great to me! Could you integrate this plugin with the Agent Client, following the same approach used for other plugin integrations? 
can you also make a PR on Develop branch

alextitonis commented:
> Looks great to me! Could you integrate this plugin with the Agent Client, following the same approach used for other plugin integrations? can you also make a PR on Develop branch

Tried to follow the other plugins, is there something wrong? can fix it. 
Will do the other PR as well

samarth30 commented:
Hi @alextitonis,

Could you include the plugin initialization in agent/src/index.ts, similar to how the other plugins have been integrated? This will ensure consistency across the project.

Also, it might be a good idea to define some actions so your plugin can be utilized effectively. By attaching the plugin to specific actions, it can seamlessly integrate into the workflow. Let me know if you’d like any support with this!

alextitonis commented:
> Hi @alextitonis,
> 
> Could you include the plugin initialization in agent/src/index.ts, similar to how the other plugins have been integrated? This will ensure consistency across the project.
> 
> Also, it might be a good idea to define some actions so your plugin can be utilized effectively. By attaching the plugin to specific actions, it can seamlessly integrate into the workflow. Let me know if you’d like any support with this!

Rebased to developer and added initialization, would be perfect if you could help with the action example

wtfsayo commented:
>would be perfect if you could help with the action example

https://github.com/elizaOS/eliza/blob/8205313071203e2610c9817839503f60919f7a40/packages/plugin-image-generation/src/index.ts#L65

wtfsayo reviewed (changes requested):
Please add actions to make it work!

alextitonis commented:
@wtfsayo @samarth30 added an action

wtfsayo commented:
@coderabbitai review
